HotTubController/HotTubController.ino

```diff
@@ -29,62 +31,145 @@ enum HeatingMode {
+// where as an alpha of 0.0001f takes about 30 seconds for a similar response. 
+String _fileCompiledInfo;
 
 bool _isSleep = false;
 bool _deadManSwitch = true;
 HeatingMode _heatingStatus = NEITHER;
 
+// enable soft reset
+void(* resetFunc) (void) = 0;
+
 void setup(void) {
-  Serial.begin(9600);
+  // Check serial rates at: https://wormfood.net/avrbaudcalc.php
+  // Uno typically has a 16Mhz crystal, could use conditional compilation arguments here to optimize for specific boards.
+  Serial.begin(250000);  //Serial.begin(9600); 
   pinMode(HEATERPIN, OUTPUT);
   pinMode(SAFTEYPIN, OUTPUT);
   pinMode(SLEEPSWITCH, INPUT_PULLUP);
+  _fileCompiledInfo = outFileCompiledInfo();
 }
 
-
 void loop(void) {
-
-  float reading = analogRead(THERMISTORPIN);
-  
-  _isSleep = digitalRead(SLEEPSWITCH)?;
-  
-  float mesuredResistance = CalculateResistance(reading);
-
+  //Allways execute
   //TODO: Add checks against DeadMansSwitch
-  SafteyCheck(mesuredResistance);
+  float reading = analogRead(THERMISTORPIN);
+  //TODO: #3 handle common Thermistor Read errors 
+  //  - reading > 1018 :A a detached thermistor, but SERIESRESISTOR is still in place
+  //  - reading < 7 : thermistor is grounded or A0 is detached
+  //  - readings that frequently cycle from 0 through 1023 (oddly around 60 hz durring testing) 
+  //    occures when A0 is connected, but not powered or grounded.
+  float measuredResistance = CalculateResistance(reading);
+
+   // Calculate the EMA of the measured resistance.
+  _emaResistance = CalculateExponentialMovingAverage(_alpha,_emaResistance, measuredResistance);
+  // _emaSafteyResistance is much more responsive than _emaResistance
+  _emaSafteyResistance = CalculateExponentialMovingAverage(_alphaSaftey,_emaSafteyResistance, measuredResistance);
+ 
+  SafteyCheck(_emaSafteyResistance);
+  //unsigned long currentRunTime = millis(); 
   
-  if(_deadManSwitch = false){
+  //for testing time overflows by using micros() as it oveflows in 70 minutes   
+  unsigned long currentRunTime = micros();
   
-  float targetHi;
-  float targetLow;
-  OutGetTargetTemp(targetHi, targetLow);
-
-  SetHeatingStatus(mesuredResistance, targetHi, targetLow);
-
-  SetHeater();
-
+  // The number of milliseconds since board's last reset
+  // Unsigned Long is 32 bit and overflows after 4,294,967,295  (2^32-1)
+  // millis overflows ever 49.8 days
+  // an unsigned negitive value is a positive value
+  if ((unsigned long)(currentRunTime - _previousRunTime) > (ACTION_INTERVAL-1)) {
+    // only take actions if ACTION_INTERVAL has passed
+    if (_deadManSwitch = false) {
+      float targetHi;
+      float targetLow;
+      OutGetTargetTemp(targetHi, targetLow);
+      SetHeatingStatus(targetHi, targetLow);
+      SetHeater();
+    }
+    
+    // Build Log Message as json-logs:https://signoz.io/blog/json-logs/
+    // e.g.->  {"RunTime":1235,}
+    String logOuput = "{";
+    logOuput.concat("\"BootTime\":");
```

Comment:
Can we move this into a reporting function?

Reply:
I've just started out report on this next commit, if you can figure out how to get this information once, instead of calling it everytime, just do it in outReport().

Reply:
Does it make more sense to have multiple report files, or continue with the current flat json structure, or nested objects like:
[{
    "board":"ID",
    "messageType":"Info"
    "message":[{
        "attributes":"values"
    }]
}] \n
[{
    "board":"ID",
    "messageType":"Error"
    "debuggingLevel":1
    "message":"error message"
}] \n 
...

 I'd prefer one of the last two, just use the FileCompiledInfo to identify multiple boards, not really relevent for this project, but it would be nice to just use that as the quasi unique id of the arduino board, it tells you down to the second the time the sketch was compiled, so as long as you're not using a programmer each board will have a unique value. 

I started using https://reference.arduino.cc/reference/en/libraries/jsonlogger/ but it consumed 20% of available programming space.

Reply:
Board ID as a hash of FileCompiledInfo?  https://github.com/esp8266/Arduino/blob/master/libraries/Hash/src/Hash.h